src/Aspire.Hosting/Dashboard/proto/resource_service.proto

```diff
@@ -0,0 +1,190 @@
+}
+
+message ApplicationInformationResponse {
+    string application_name = 1;
+    string application_version = 2;
+}
+
+////////////////////////////////////////////
+
+message ResourceCommandRequest {
+    string command_type = 1;
+    ResourceId resource_id = 2;
+    // When present, this message must be shown to the user and their confirmation obtained
+    // before sending the request for this command to be executed.
+    // The user will be presented with Ok/Cancel options.
+    optional string confirmation_message = 3;
+    // Optional parameter that configures the command in some way.
+    // Clients must return any value provided by the server when invoking
+    // the command.
+    optional google.protobuf.Value parameter = 4;
+}
+
+enum ResourceCommandResponseKind {
+    UNDEFINED = 0;
+    SUCCEEDED = 1;
+    FAILED = 2;
+    CANCELLED = 3;
+}
+
+message ResourceCommandResponse {
+    ResourceCommandResponseKind kind = 1;
+    optional string error_message = 2;
+}
+
+////////////////////////////////////////////
+
+message ResourceType {
+    // Unique name for the resource type. Equivalent to ResourceId.resource_type
+    // If "display_name" is omitted, this value will be used in UIs.
+    string unique_name = 1;
+
+    // Display string for references to this type in UI. May be localized.
+    // If this value is omitted, UIs will show "unique_name" instead.
+    optional string display_name = 2;
+
+    // Any commands that may be executed against resources of this type, avoiding
+    // the need to copy the value to every Resource instance.
+    //
+    // Note that these commands must apply to matching resources at any time.
+    //
+    // If the set of commands changes over time, use the "commands" property
+    // of the Resource itself.
+    repeated ResourceCommandRequest commands = 3;
+}
+
+////////////////////////////////////////////
+
+message EnvironmentVariable {
+    string name = 1;
+    optional string value = 2;
+    optional bool is_from_spec = 3;
+}
+
+message Endpoint {
+    string name = 1;
+    optional string allocated_address = 2;
+    optional int32 allocated_port = 3;
+    optional string http_address = 4;
+}
+
+message AdditionalData {
+    // Name of the data item, e.g. "id", "image", "path", ...
+    string name = 1;
+    // Optional namespace, e.g. "container", "executable", "project", ...
+    optional string namespace = 2;
+    // Optional display name, may be localized
+    optional string display_name = 3;
+    // The data value. May be null, a number, a string, a boolean, a dictionary of values (Struct), or a list of values (ValueList).
+    google.protobuf.Value value = 4;
+}
+
+message ResourceId {
+    string uid = 1;
+    // TODO do we need resource_type to make unique names? if not, inline ResourceId type as string.
+    string resource_type = 2;
+}
+
+// Models the full state of an resource (container, executable, project, etc) at a particular point in time.
+message ResourceSnapshot {
```

Comment:
Could `ResourceSnapshot` just be `Resource`? I think it's implicit that all the data returned from a service is a snapshot in time. Adding snapshot to the message name is unnecessary.

Reply:
Perhaps. I try to avoid very generic names like "resource" as they tend to conflict. That said, we'll only be using these messages in two places in C#, as we'll only create/consume gRPC types at the edges, and convert to/from our own classes beyond that.